packages/core/src/render3/i18n.ts

```diff
@@ -932,68 +950,98 @@ function readUpdateOpCodes(
   }
 }
 
-function icuUpdateCase(
-    tView: TView, lView: LView, tIcuIndex: number, nodeIndex: number, bindingsStartIndex: number,
-    tIcus: TIcu[], caseCreated: boolean) {
-  const tIcu = tIcus[tIcuIndex];
-  const icuTNode = getTNode(tView, nodeIndex) as TIcuContainerNode;
-  if (icuTNode.activeCaseIndex !== null) {
-    readUpdateOpCodes(
-        tIcu.update[icuTNode.activeCaseIndex], tIcus, bindingsStartIndex, changeMask, tView, lView,
-        caseCreated);
+/**
+ * Apply OpCodes associated with updating an existing ICU.
+ *
+ * @param tView Current `TView`
+ * @param tIcus If ICUs present than this contain them.
+ * @param icuIndex Index into `tIcus` to process.
+ * @param bindingsStartIndex Location of the first `ɵɵi18nApply`
+ * @param lView Current `LView`
+ * @param changeMask Each bit corresponds to a `ɵɵi18nExp` (Counting backwards from
+ *     `bindingsStartIndex`)
+ */
+function applyIcuUpdateCase(
+    tView: TView, tIcus: TIcu[], icuIndex: number, bindingsStartIndex: number, lView: LView,
+    caseCreated: boolean) {
+  ngDevMode && assertIndexInRange(tIcus, icuIndex);
+  const tIcu = tIcus[icuIndex];
+  ngDevMode && assertIndexInRange(lView, tIcu.currentCaseLViewIndex);
+  const activeCaseIndex = lView[tIcu.currentCaseLViewIndex];
+  if (activeCaseIndex !== null) {
+    const mask = caseCreated ?
+        -1 :  // -1 is same as all bits on, which simulates creation since it marks all bits dirty
+        changeMask;
+    applyUpdateOpCodes(tView, tIcus, lView, tIcu.update[activeCaseIndex], bindingsStartIndex, mask);
   }
 }
 
-function icuSwitchCase(
-    tView: TView, tIcuIndex: number, nodeIndex: number, tIcus: TIcu[], lView: LView,
-    value: string): boolean {
-  const tIcu = tIcus[tIcuIndex];
-  const icuTNode = getTNode(tView, nodeIndex) as TIcuContainerNode;
+/**
+ * Apply OpCodes associated with switching a case on ICU.
+ *
+ * This involves tearing down existing case and than building up a new case.
+ *
+ * @param tView Current `TView`
+ * @param tIcus ICUs active at this location.
+ * @param icuIndex Index into `tIcus` to process.
+ * @param lView Current `LView`
+ * @param value Value of the case to update to.
+ * @returns true if a new case was created (needed so that the update executes regardless of the
+ *     bitmask)
+ */
+function applyIcuSwitchCase(
+    tView: TView, tIcus: TIcu[], icuIndex: number, lView: LView, value: string): boolean {
+  applyIcuSwitchCaseRemove(tView, tIcus, icuIndex, lView);
+
+  // Rebuild a new case for this ICU
   let caseCreated = false;
-  // If there is an active case, delete the old nodes
-  if (icuTNode.activeCaseIndex !== null) {
-    const removeCodes = tIcu.remove[icuTNode.activeCaseIndex];
+  const tIcu = tIcus[icuIndex];
```

Comment:
May be add the checks that the index is in the range here as well?
```suggestion
  ngDevMode && assertIndexInRange(tIcus, icuIndex);
  const tIcu = tIcus[icuIndex];
```

Reply:
done